File: android/app/src/main/kotlin/dev/agentdeck/terrarium/environment/WaterEffect.kt
package dev.agentdeck.terrarium.environment

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.BlendMode
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.DrawScope
import dev.agentdeck.terrarium.EnvironmentVisualState
import dev.agentdeck.terrarium.TerrariumColors
import dev.agentdeck.terrarium.TerrariumLayout
import dev.agentdeck.terrarium.TerrariumTiming
import kotlin.math.PI
import kotlin.math.sin

/**
 * Caustics light pattern — overlapping sine meshes drawn with overlay blend.
 * Intensity varies with environment state.
 */
class WaterEffect {

    private var envState by mutableStateOf(EnvironmentVisualState.CALM)
    private var time by mutableFloatStateOf(0f)
    private var surfaceTime by mutableFloatStateOf(0f)

    fun setState(state: EnvironmentVisualState) {
        envState = state
    }

    fun update(dt: Float) {
        time += dt * TerrariumTiming.CAUSTICS_SPEED
        val speedMul = when (envState) {
            EnvironmentVisualState.DARK -> 0.3f
            EnvironmentVisualState.CALM -> 1.0f
            EnvironmentVisualState.ACTIVE -> 1.6f
            EnvironmentVisualState.ALERT -> 1.1f
        }
        surfaceTime += dt * TerrariumTiming.SURFACE_WAVE_SPEED * speedMul
    }

    /**
     * Draw animated water surface — filled wave regions create air/water contrast.
     *
     * Instead of thin stroke lines (invisible), we fill the area ABOVE the wave curve
     * with a lighter tint. The surface is perceived as the boundary between two regions,
     * not as a drawn line.
     */
    fun drawSurface(scope: DrawScope) {
        val w = scope.size.width
        val h = scope.size.height
        val surfaceY = h * TerrariumLayout.WATER_SURFACE_Y_FRACTION

        // Amplitude relative to canvas — visible at any resolution
        val amp = h * when (envState) {
            EnvironmentVisualState.DARK -> 0.003f
            EnvironmentVisualState.CALM -> 0.008f
            EnvironmentVisualState.ACTIVE -> 0.014f
            EnvironmentVisualState.ALERT -> 0.009f
        }
        val fillAlpha = when (envState) {
            EnvironmentVisualState.DARK -> 0.03f
            EnvironmentVisualState.CALM -> 0.08f
            EnvironmentVisualState.ACTIVE -> 0.12f
            EnvironmentVisualState.ALERT -> 0.09f
        }

        val twoPi = 2f * PI.toFloat()

        // (1) Primary wave — filled from curve up to top of canvas
        //     Creates the main air/water tonal boundary
        drawFilledWave(scope, w, surfaceY, amp,
            freq = twoPi / (w * 0.6f),
            phase = surfaceTime * twoPi,
            fillAlpha = fillAlpha)

        // (2) Secondary wave — shorter wavelength, smaller amplitude, opposite direction
        //     Overlaps with primary to create natural interference shimmer
        drawFilledWave(scope, w, surfaceY, amp * 0.4f,
            freq = twoPi / (w * 0.35f),
            phase = -surfaceTime * twoPi * 1.4f + 1.5f,
            fillAlpha = fillAlpha * 0.5f)

        // (3) Sub-surface glow — bright gradient just below the wave line
        //     Simulates light refraction at water surface
        val glowDepth = h * 0.025f
        scope.drawRect(
            brush = Brush.verticalGradient(
                colors = listOf(
                    Color.White.copy(alpha = fillAlpha * 0.7f),
                    Color.Transparent,
                ),
                startY = surfaceY,
                endY = surfaceY + glowDepth,
            ),
            topLeft = Offset(0f, surfaceY),
            size = Size(w, glowDepth),
        )
    }

    /**
     * Fill the region from a sine wave curve up to y=0 (top of canvas).
     * The filled area represents "air" — slightly brighter than water below.
     */
    private fun drawFilledWave(
        scope: DrawScope, w: Float, baseY: Float, amplitude: Float,
        freq: Float, phase: Float, fillAlpha: Float,
    ) {
        val path = Path().apply {
            // Start at top-left corner
            moveTo(0f, 0f)

            // Walk the wave curve left to right
            val step = 3f
            var x = 0f
            while (x <= w) {
                val y = baseY + sin(freq * x + phase) * amplitude
                lineTo(x, y)
                x += step
            }
            // Ensure we reach the right edge
            lineTo(w, baseY + sin(freq * w + phase) * amplitude)

            // Close back to top-right → top-left
            lineTo(w, 0f)
            close()
        }
        scope.drawPath(
            path = path,
            color = Color.White.copy(alpha = fillAlpha),
        )
    }

    fun draw(scope: DrawScope) {
        if (envState == EnvironmentVisualState.DARK) return

        val w = scope.size.width
        val h = scope.size.height

        val alpha = when (envState) {
            EnvironmentVisualState.DARK -> 0f
            EnvironmentVisualState.CALM -> 0.08f
            EnvironmentVisualState.ACTIVE -> 0.12f
            EnvironmentVisualState.ALERT -> 0.10f
        }

        // Draw two overlapping caustic layers with different phases
        drawCausticLayer(scope, w, h, alpha, phase = 0f)
        drawCausticLayer(scope, w, h, alpha * 0.6f, phase = PI.toFloat() * 0.7f)
    }

    private fun drawCausticLayer(
        scope: DrawScope, w: Float, h: Float, alpha: Float, phase: Float,
    ) {
        val cellSize = w / GRID_SIZE

        for (row in 0 until GRID_SIZE) {
            for (col in 0 until GRID_SIZE) {
                val baseX = col * cellSize
                val baseY = row * cellSize

                // Distorted diamond shape
                val offset1 = sin(time + col * 0.5f + phase) * cellSize * 0.3f
                val offset2 = sin(time * 0.7f + row * 0.4f + phase) * cellSize * 0.3f

                val cx = baseX + cellSize / 2 + offset1
                val cy = baseY + cellSize / 2 + offset2

                val size = cellSize * (0.3f + sin(time * 1.3f + col + row + phase) * 0.15f)

                val path = Path().apply {
                    moveTo(cx, cy - size)
                    lineTo(cx + size * 0.8f, cy)
                    lineTo(cx, cy + size)
                    lineTo(cx - size * 0.8f, cy)
                    close()
                }

                scope.drawPath(
                    path = path,
                    color = TerrariumColors.CausticsLight.copy(alpha = alpha),
                    blendMode = BlendMode.Overlay,
                )
            }
        }
    }

    companion object {
        private const val GRID_SIZE = 8
    }
}

File: apple/AgentDeck/Terrarium/Environment/WaterEffect.swift
// WaterEffect.swift — Caustics light pattern with overlapping sine meshes
// Ported from android WaterEffect.kt

import SwiftUI

final class WaterEffect {
    private var time: Float = 0
    private var envState: EnvironmentVisualState = .calm

    private static let lineCount = 5

    func setState(_ state: EnvironmentVisualState) {
        envState = state
    }

    func update(dt: Float) {
        time += dt * TerrariumTiming.causticsSpeed
    }

    func draw(context: inout GraphicsContext, size: CGSize) {
        guard envState != .dark else { return }

        let w = Float(size.width)
        let h = Float(size.height)

        let alpha: Float = switch envState {
        case .dark: 0
        case .calm: 0.08
        case .active: 0.12
        case .alert: 0.10
        }

        // Use additive blending (matching Android BlendMode.Plus)
        // This creates a lighter, more ethereal caustic effect instead of opaque overlay
        var additive = context
        additive.blendMode = .plusLighter

        // Two overlapping caustic layers with different phases
        drawCausticLayer(context: &additive, w: w, h: h, alpha: alpha, phase: 0)
        drawCausticLayer(context: &additive, w: w, h: h, alpha: alpha * 0.05, phase: Float.pi * 0.7)
    }

    private func drawCausticLayer(context: inout GraphicsContext, w: Float, h: Float,
                                   alpha: Float, phase: Float) {
        let twoPi: Float = 2 * .pi
        let spacing = w / Float(Self.lineCount)
        let waveLen1 = w * 0.4
        let waveLen2 = w * 0.32
        let amp = spacing * 0.35
        let strokeW = CGFloat(w * 0.004)
        let reducedAlpha = alpha * 0.04

        let freq1 = twoPi / waveLen1
        let freq2 = twoPi / waveLen2
        let step: Float = 6

        // Family 1: near-horizontal lines (~10° tilt)
        let angle1: Float = 10 * .pi / 180
        let sin1 = sin(angle1), cos1 = cos(angle1)
        let extent = w * 0.15

        for i in 0..<Self.lineCount {
            let lineOffset = (Float(i) - Float(Self.lineCount) / 2) * spacing
            let linePhase = phase + Float(i) * 0.7
            var path = Path()
            var t: Float = -extent
            var first = true
            while t <= w + extent {
                let wave = sin(freq1 * t + time + linePhase) * amp
                let x = t * cos1 - (lineOffset + wave) * sin1
                let y = t * sin1 + (lineOffset + wave) * cos1 + h * 0.5
                let pt = CGPoint(x: CGFloat(x), y: CGFloat(y))
                if first { path.move(to: pt); first = false } else { path.addLine(to: pt) }
                t += step
            }
            context.stroke(path,
                           with: .color(.white.opacity(Double(reducedAlpha))),
                           lineWidth: strokeW)
        }

        // Family 2: ~60° angled lines
        let angle2: Float = 60 * .pi / 180
        let sin2 = sin(angle2), cos2 = cos(angle2)
        let diag = w + h

        for i in 0..<Self.lineCount {
            let lineOffset = (Float(i) - Float(Self.lineCount) / 2) * spacing * 1.2
            let linePhase = phase + Float(i) * 0.9 + 2.0
            var path = Path()
            var t: Float = -extent
            var first = true
            while t <= diag + extent {
                let wave = sin(freq2 * t + time * 0.85 + linePhase) * amp
                let x = t * cos2 - (lineOffset + wave) * sin2
                let y = t * sin2 + (lineOffset + wave) * cos2
                let pt = CGPoint(x: CGFloat(x), y: CGFloat(y))
                if first { path.move(to: pt); first = false } else { path.addLine(to: pt) }
                t += step
            }
            context.stroke(path,
                           with: .color(.white.opacity(Double(reducedAlpha))),
                           lineWidth: strokeW)
        }
    }
}
